[MM-26478] Add one click reactions

isacikgoz opened the pull request:
#### Summary

Adds one-click reactions to the posts. Can be configured. This is my first ever frontend feature, so it probably have massive errors in it :) I'd like to pair with someone to go over the changes and polish the PR. Therefore I'm opening this as a draft PR.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-26478

#### Screenshots

![ezgif com-gif-maker](https://user-images.githubusercontent.com/2153367/133612510-57934241-6e60-4756-b8eb-8ef8b932c863.gif)


#### Release Note

```release-note
Added one-click reactions for the posts. The 3 most recently used emojis will show up when the mouse is hovered on a post. 
```

esethna reviewed (changes requested):
AWESOME! Few comments:

1. Shouldn't the ... be on the far right?

![image](https://user-images.githubusercontent.com/14333569/133674744-16835b48-c1e7-4948-bc93-9337505ff78c.png)

2. I'll let @Rina-dsg comment on the size and alignment of the emojis in the hover menu but they look a bit too small and slightly misaligned

3. @Rina-dsg should we consider updating this emoji icon to the face with a little plus on it to indicate that there are "more" emoji reactions behind that button opposed to it being the only entry point for reactions in the past? 
![image](https://user-images.githubusercontent.com/14333569/133675158-3b596da6-cea7-4854-809e-5a31535d8e4f.png)

Otherwise this works beautifully

Rina-dsg commented:
1. @isacikgoz do emojis are the same size as in the figma file? as they seem a bit smaller
2. @esethna Good idea with updating the emoji icon. Updated the figma file
3. More option (3 dots icon) in the hover menu should be positioned on the right
<img width="527" alt="Screenshot 2021-09-16 at 23 55 48" src="https://user-images.githubusercontent.com/85259678/133684468-0d9cb30f-08fa-4ac2-964e-c55b86fe3c31.png">

isacikgoz commented:
Thanks for the feedback @esethna and @Rina-dsg much appreciated, addressed the emoji size issue and updated the emoji icon. 

<img width="245" alt="image" src="https://user-images.githubusercontent.com/2153367/133764887-0e28475f-7410-45db-a99e-eb5725444c7b.png">

The 3 dot button position is decided by the collapsed reply threads feature, If it is enabled the button is rendered on the right hand side. If not, on the left hand side. Shall we make it on the right hand side by default?

Rina-dsg commented:
@isacikgoz emoji reactions look good now 🙂. Regarding the 3 dots icon, I think we should make this icon on the right by default. It's better to have a consistent position of the icon.

Rina-dsg commented:
@isacikgoz just noticed, for 👍 -> should be 'thumbsup' on the tooltip

esethna reviewed (changes requested):
@isacikgoz ~~I think the ... should be on the right regardless of CRT?~~ 

Edit: You're correct, if CRT is off then the ... previously showed on the left since

esethna commented:
> just noticed, for 👍 -> should be 'thumbsup' on the tooltip

@rina I think it's fine if the tooltip is +1, it's just based on it's default emoji string. Also GitHub uses "+1" so it's familiar to our target audience of devs

esethna commented:
Heads up @jfrerich hopefully there aren't too many conflicts with your work in this PR

jfrerich commented:
@esethna, at first glance, it doesn't look like it will be too bad. The components we both touched are changing different parts of the code.

esethna commented:
Thanks, looks great as is! Edited my comment above as I was mistaken

koox00 commented:
> @isacikgoz emoji reactions look good now 🙂. Regarding the 3 dots icon, I think we should make this icon on the right by default. It's better to have a consistent position of the icon.

Just to remind you the 3 dots menu is not on the right when CRT is off because the replies button is on the right.
I don't think we should change that. 
The replies button is rendered even without hovering, so upon hovering I'd expect it to be in the same place.

koox00 commented on components/post_view/post_recent_reactions/recent_reactions_emoji_item.jsx:
```diff
@@ -0,0 +1,46 @@
+        onItemClick: PropTypes.func.isRequired,
+    };
+
+    emojiItemRef = (emojiItem) => {
+        this.emojiItem = emojiItem;
+    };
```
nit but since we don't use the ref we could as well delete it
```suggestion
```

koox00 commented on components/post_view/post_recent_reactions/recent_reactions_emoji_item.jsx:
```diff
@@ -0,0 +1,46 @@
+        return (
+            <div
+                className={itemClassName}
+                ref={this.emojiItemRef}
```
```suggestion
```

koox00 commented on components/post_view/reaction/reaction.scss:
```diff
@@ -66,6 +66,14 @@
     }
 
+    &__emoji--post-menu {
+        min-width: 18px;
+        max-width: 18px;
+        min-height: 18px;
+        max-height: 18px;
```
nit: I guess we could simplify by just:
```suggestion
        width: 18px;
        height: 18px;
```

koox00 commented on components/post_view/post_info/post_info.tsx:
```diff
@@ -210,6 +215,27 @@ export default class PostInfo extends React.PureComponent<Props, State> {
+        const showRecentlyUsedReactions = !isSystemMessage && hover && !isReadOnly && this.props.oneClickReactionsEnabled;
+        let showRecentReacions;
+        if (showRecentlyUsedReactions) {
+            const state = store.getState();
+            const recentEmojis = getRecentEmojis(state).slice(-3);
+            const emojiMap = getEmojiMap(state);
+            const emojis = [];
+            for (let i = 0; i < recentEmojis.length; i++) {
+                emojis.push(emojiMap.get(recentEmojis[i]));
+            }
```
same as last comment

koox00 commented on components/post_view/post_recent_reactions/post_recent_reactions.tsx:
```diff
@@ -0,0 +1,125 @@
+    showEmojiPicker: boolean;
+}
+
+const state = store.getState();
+const emojiMap = getEmojiMap(state);
```
This too would belong in a `mapStateToProps` function in the component's index file.

koox00 commented on components/rhs_comment/rhs_comment.jsx:
```diff
@@ -449,6 +454,27 @@ export default class RhsComment extends React.PureComponent {
+        const showRecentlyUsedReactions = (!isReadOnly && !isEphemeral && !post.failed && !isSystemMessage && !channelIsArchived && this.props.oneClickReactionsEnabled);
+        let showRecentReacions;
+        if (showRecentlyUsedReactions) {
+            const state = store.getState();
+            const recentEmojis = getRecentEmojis(state).slice(-3);
+            const emojiMap = getEmojiMap(state);
+            const emojis = [];
+            for (let i = 0; i < recentEmojis.length; i++) {
+                emojis.push(emojiMap.get(recentEmojis[i]));
+            }
```
Same as last comment

koox00 commented on components/rhs_root_post/rhs_root_post.jsx:
```diff
@@ -235,6 +239,27 @@ export default class RhsRootPost extends React.PureComponent {
+        const showRecentlyUsedReactions = (!isReadOnly && !isEphemeral && !post.failed && !isSystemMessage && !channelIsArchived && this.props.oneClickReactionsEnabled);
+        let showRecentReacions;
+        if (showRecentlyUsedReactions) {
+            const state = store.getState();
+            const recentEmojis = getRecentEmojis(state).slice(-3);
+            const emojiMap = getEmojiMap(state);
+            const emojis = [];
+            for (let i = 0; i < recentEmojis.length; i++) {
+                emojis.push(emojiMap.get(recentEmojis[i]));
+            }
```
we should only consume redux store in a connect function for class components.
This is usually located in the component's `index` file.

You could also write a selector to re-use:

e.g:
```ts
const getNamingIsHardEmojis = (state) => {
    const recentEmojis = getRecentEmojis(state).slice(-3);
    const emojiMap = getEmojiMap(state);
    const emojis = [];
    for (let i = 0; i < recentEmojis.length; i++) {
        emojis.push(emojiMap.get(recentEmojis[i]));
    }
    
    return emojis;
}
```

koox00 reviewed (changes requested):
Looks great 💪 !! Added some comments on how to use store, and some nitpicks, other than that I'm excited to have this feature in!!

koox00 commented on components/post_view/post_recent_reactions/post_recent_reactions.tsx:
```diff
@@ -0,0 +1,118 @@
+        this.props.actions.addReaction(this.props.postId, emojiName);
+    };
+
+    complementEmojis(): void {
+        const additional = this.props.defaultEmojis.filter((e) => {
+            let ignore = false;
+            for (const emoji of this.props.emojis) {
+                if (e.name === emoji.name) {
+                    ignore = true;
+                    break;
+                }
+            }
+            return !ignore;
+        });
+        const l = this.props.emojis.length;
+        for (let i = 0; i < 3 - l; i++) {
+            this.props.emojis.push(additional[i]);
+        }
+    }
```
we shouldn't mutate props directly, see [this](https://reactjs.org/blog/2015/02/24/streamlining-react-elements.html) for more info.
If it's possible to compliment in the selector that would be best so we avoid duplication as well

koox00 reviewed (approved):
💪

isacikgoz commented:
Thanks a lot for the thorough review @koox00 it helped a lot! 💯

esethna commented:
@isacikgoz one thing to consider here if not already, there is an admin setting for the emoji picker. So if that's disabled, we shouldn't show these: https://docs.mattermost.com/configure/configuration-settings.html#enable-emoji-picker

```
"EnableEmojiPicker": true

jgilliam17 commented:
Thanks @isacikgoz Love this option to add one click reactions to the post.

Few things I noticed:

1. Checked the use case when system console setting for emoji picker is set to False, like Eric mentioned and the account setting for one click reactions is not removed. We should hide it.
Additionally, sysadmin still had 3 emoji option on post hover when account setting is ON and  emoji picker is set to false. Not sure how that is happening.

2. On RHS all 3 are shown on hover, per ticket we should show 1 reaction on minimized RHS and 3 when RHS is expanded
<img width="1199" alt="Screen Shot 2021-09-22 at 4 01 33 PM" src="https://user-images.githubusercontent.com/52937121/134413558-0825e89a-5380-4210-97f3-8a8d33cddee9.png">
<img width="1204" alt="Screen Shot 2021-09-22 at 4 01 21 PM" src="https://user-images.githubusercontent.com/52937121/134413565-9a9dd8a6-ae17-4849-bec9-f77c16984329.png">

The following 2 points are on mobile view and can be filed separately if necessary as they are not part of the original design. Needs UX input on how to handle one-click reactions on mobile view @Rina-dsg 

3. On mobile view, when viewing RHS, one click reactions are always visible, no need to hover. 
<img width="743" alt="Screen Shot 2021-09-22 at 4 02 32 PM" src="https://user-images.githubusercontent.com/52937121/134413885-53c2f5e0-0de2-470f-8166-d5e379aa7d7f.png">

4. On mobile view with CRT off, the reply arrow switches position with 3-dots on hover.
<img width="738" alt="Screen Shot 2021-09-22 at 4 06 01 PM" src="https://user-images.githubusercontent.com/52937121/134415452-6557f868-045a-4722-a84c-255e2f4c1926.png">

calebroseland reviewed (approved):
LGTM 🎉

isacikgoz commented:
Thanks for the detailed feedback @jgilliam17 Never tested this with mobile view 🤦  I think I addressed those but please take a look when you have the time :)

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Quite a few E2Es failed. I'll have to re-run again after update.
There was an issue that was recently fixed (extra post space) that is still failing on the server making it hard to discern failures related to PR changes. 
See current results [here ](https://mm-cypress-report.s3.amazonaws.com/47731-dfa65b5-ee-ent-webapp-pr-8866-mattermost-enterprise-edition:dfa65b5/mochawesome.html)

jgilliam17 commented:
@isacikgoz Can you please update E2Es that are failing 
Latest [E2Es results](https://mm-cypress-report.s3.amazonaws.com/47750-1e46cee-ee-ent-webapp-pr-8866-mattermost-enterprise-edition:1e46cee/mochawesome.html)
Latest [master run results](https://mm-cypress-report.s3.amazonaws.com/47757-ee-ent-master-mm-ee-test:test/mochawesome.html) for comparison
After looking at the failures these two sections are related to PR changes
1. Accessibility support in Account settings is failing due to new setting option for One Click reactions - 6 tests
2. Accessibility support in post menu (unable to tab through one click reactions) on main menu and RHS - 2 tests
For the 2nd one, we may have to duplicate this test and keep a version to run with one click reaction off and one test with one click reactions on.  @saturninoabril can you please advise, my suggestion may be unnecessary. Thanks.

saturninoabril commented:
Thanks @jgilliam17, I totally agree!

isacikgoz commented:
@jgilliam17 @saturninoabril I updated the `spec` files to make tests pass. I believe the cases for accessibility support in post menu doesn't require additional tests for one click reactions as it only tabs through buttons. All I did was adding new checks to verify scrolling through buttons. On the other hand, I can add new automated cypress test cases for specifically testing the one-click reactions if you like. Whether a part for this PR or with another PR. (I will be on PTO for a week so I'd rather add test in a separate PR 😄)

saturninoabril commented:
Thanks @isacikgoz, sounds good to me.

saturninoabril commented:
Recent E2E test is now all good. No related failed test - https://mm-cypress-report.s3.amazonaws.com/47812-ee-ent-master-mm-ee-test:test/mochawesome.html

jgilliam17 reviewed (approved):
Thanks @isacikgoz 
Tested, looks good to merge.
- Verified one-click reactions feature and created a [test case for E2Es](https://mattermost.atlassian.net/projects/MM?selectedItem=com.atlassian.plugins.atlassian-connect-plugin:com.kanoah.test-manager__main-project-page#!/testCase/MM-T4261). Test case linked in a follow up task https://mattermost.atlassian.net/browse/MM-39032
